p4-16/psa/examples/psa-example-digest.p4

```diff
@@ -0,0 +1,216 @@
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        buffer.extract(parsed_hdr.ipv4);
+        transition accept;
+    }
+}
+
+parser IngressParserImpl(packet_in buffer,
+                         packed_metadata_in packed_meta,
+                         out headers parsed_hdr,
+                         inout metadata meta,
+                         in psa_ingress_parser_input_metadata_t istd,
+                         out psa_parser_output_metadata_t ostd)
+{
+    CommonParser() p;
+
+    state start {
+        transition packet_in_parsing;
+    }
+
+    state packet_in_parsing {
+        p.apply(buffer, parsed_hdr, meta);
+        transition accept;
+    }
+}
+
+parser EgressParserImpl(packet_in buffer,
+                        out headers parsed_hdr,
+                        inout metadata meta,
+                        in psa_egress_parser_input_metadata_t istd,
+                        out psa_parser_output_metadata_t ostd)
+{
+    CommonParser() p;
+
+    state start {
+        transition packet_in_parsing;
+    }
+
+    state packet_in_parsing {
+        p.apply(buffer, parsed_hdr, meta);
+        transition accept;
+    }
+}
+
+control ingress(inout headers hdr,
+                inout metadata meta,
+                PacketReplicationEngine pre,
+                in    psa_ingress_input_metadata_t  istd,
+                inout psa_ingress_output_metadata_t ostd)
+{
+
+    action do_digest () {
+        meta.digest_id = 0;
+        meta.digest_h.src_addr = hdr.srcAddr;
+        meta.digest_h.ingress_port = istd.ingress_port;
+    }
+    table mac_cache {
+        keys = {
+            hdr.srcAddr : exact;
+            meta.ingress_port : exact;
+        }
+        actions = {
+            do_digest; nop;
+        }
+        default_action = do_digest();
+    }
+    action do_switch (bit<9> egress_port) {
+        ostd.egress_port = egress_port;
+    }
+    table l2_tbl {
+        keys = {
+            hdr.dstAddr : exact;
+        }
+        actions = {
+            do_switch; nop;
+        }
+        default_action = nop();
+    }
+    apply {
+        mac_cache.apply();
+        l2_tbl.apply();
+    }
+}
+
+
+control egress(inout headers hdr,
+               inout metadata meta,
+               BufferingQueueingEngine bqe,
+               in    psa_egress_input_metadata_t  istd,
+               inout psa_egress_output_metadata_t ostd)
+{
+    apply {
+
+    }
+}
+
+
+control CommonDeparserImpl(packet_out packet, inout headers hdr) {
+    apply {
+        packet.emit(hdr.ethernet);
+        packet.emit(hdr.ipv4);
+    }
+}
+
+control IngressDeparserImpl(packet_out packet,
+                            digest_out packed_meta,
+                            inout headers hdr,
+                            in metadata meta,
+                            in psa_ingress_output_metadata_t istd)
+{
+    CommonDeparserImpl() common_deparser;
+    apply {
+        if (meta.digest_id == 0) {
```

Comment:
The main problem here is code safety. Let's extend this example for several values of meta.digest_id? What's the mechanism that ensures that (a) we pack exactly what we want for the right digest_id and (b) that the control plane unpacks exactly the same thing?

I think we need to compare this to P4_14, where each field_list is a named object and thus each digest gets a separate API, based on the field_list that was used. Here it is all by number instead and thus is very error-prone...

Reply:
@vgurevich I believe this is the best PSA can do at the moment until we sort out field references in the language. We discussed that at the last LDWG and it is a topic that shall be addressed in the near future.